File: typescript/packages/host/src/ui/framed.ts
import type { EventOf, KnownEvent } from "@threads/core/host";

// The committed events a UI stream shows (spec/schema/ui/README.md, "Mapping"). Every other
// event maps to zero frames.

export type Framed = EventOf<
  | "model_request"
  | "model_response"
  | "model_response_recovered"
  | "model_attempt_abandoned"
  | "approval_requested"
  | "approval_granted"
  | "approval_denied"
  | "tool_result"
  | "tool_result_late"
  | "agent_spawned"
  | "agent_finished"
  | "retry_scheduled"
>;

const FRAMED = {
  model_request: true,
  model_response: true,
  model_response_recovered: true,
  model_attempt_abandoned: true,
  approval_requested: true,
  approval_granted: true,
  approval_denied: true,
  tool_result: true,
  tool_result_late: true,
  agent_spawned: true,
  agent_finished: true,
  retry_scheduled: true,
} satisfies Record<Framed["type"], true>;

export function isFramed(e: KnownEvent): e is Framed {
  return Object.hasOwn(FRAMED, e.type);
}

/** A model step's events: its request, then its response or abandonment. */
export type ModelEvent = EventOf<
  | "model_request"
  | "model_response"
  | "model_response_recovered"
  | "model_attempt_abandoned"
>;

type Response = EventOf<"model_response" | "model_response_recovered">;

/** A text or reasoning part's id: the model_request's event id and the part's index. */
export function partId(requestId: string, index: number): string {
  return `${requestId}:${index}`;
}

/** Each response part that shows, with its index in the response's content. */
export function shownParts(e: Response): readonly Shown[] {
  return e.data.content.flatMap((part, index): Shown[] => {
    if (part.type === "text")
      return part.text === "" ? [] : [{ kind: "text", index, text: part.text }];
    if (part.type === "reasoning")
      return part.summary === undefined || part.summary === ""
        ? []
        : [{ kind: "reasoning", index, text: part.summary }];
    if (part.type === "tool_use")
      return [
        {
          kind: "tool",
          index,
          callId: part.call_id,
          name: part.name,
          input: part.input,
        },
      ];
    return [];
  });
}

export type Shown =
  | {
      readonly kind: "text" | "reasoning";
      readonly index: number;
      readonly text: string;
    }
  | {
      readonly kind: "tool";
      readonly index: number;
      readonly callId: string;
      readonly name: string;
      readonly input: EventOf<"tool_call">["data"]["input"];
    };

File: python/src/threads/host/ui/framed.py
"""The parts of a model response a UI shows (spec/schema/ui/README.md, "Mapping"): text with
text, reasoning with a summary, and calls, each with its index in the response's content."""

from dataclasses import dataclass
from typing import Literal

from pydantic.experimental.missing_sentinel import MISSING

from threads.log import (
    JsonObject,
    ModelResponseEvent,
    ModelResponseRecoveredEvent,
    ReasoningPart,
    TextPart,
    ToolUsePart,
)


@dataclass(frozen=True, slots=True)
class Written:
    """A text or reasoning part: its text, or its reasoning summary."""

    kind: Literal["text", "reasoning"]
    index: int
    text: str


@dataclass(frozen=True, slots=True)
class Called:
    index: int
    call_id: str
    name: str
    input: JsonObject


type Shown = Written | Called


def part_id(request_id: str, index: int) -> str:
    """A text or reasoning part's id: the model_request's event id and the part's index."""
    return f"{request_id}:{index}"


def shown_parts(e: ModelResponseEvent | ModelResponseRecoveredEvent) -> list[Shown]:
    out: list[Shown] = []
    for index, part in enumerate(e.data.content):
        match part:
            case TextPart(text=text) if text:
                out.append(Written("text", index, text))
            case ReasoningPart(summary=summary) if summary is not MISSING and summary:
                out.append(Written("reasoning", index, summary))
            case ToolUsePart(call_id=call_id, name=name, input=input):
                out.append(Called(index, call_id, name, input))
            case _:
                pass
    return out
